ingestors/drupal.go

```diff
@@ -0,0 +1,125 @@
+package ingestors
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/PuerkitoBio/goquery"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/librariesio/depper/data"
+	"github.com/mmcdole/gofeed"
+)
+
+const drupalSchedule = "*/5 * * * *"
+const drupalModulesUrl = "https://www.drupal.org/project/project_module?page=%d&solrsort=ds_project_latest_release+desc"
+const drupalReleasesUrl = "https://www.drupal.org/node/%s/release/feed"
+
+type Drupal struct {
+	LatestRun time.Time
+}
+
+func NewDrupal() *Drupal {
+	return &Drupal{}
+}
+
+func (ingestor *Drupal) Schedule() string {
+	return drupalSchedule
+}
+
+func (ingestor *Drupal) Name() string {
+	return "drupal"
+}
+
+func (ingestor *Drupal) Ingest() []data.PackageVersion {
+	var results []data.PackageVersion
+
+	// Until we save LatestRun state, we need to set a LatestRun to avoid scanning every single release in the index.
+	if ingestor.LatestRun.IsZero() {
+		ingestor.LatestRun = time.Now().Add(-30 * 24 * time.Hour)
+	}
+
+	page := 0
+	done := false
+	// 100 is an arbitrary limit to ensure we don't scrape all ~2k pages of packages
+	for page < 100 && !done {
+		doc, err := getHtmlDocument(fmt.Sprintf(drupalModulesUrl, page))
+		if err != nil {
+			log.WithFields(log.Fields{"ingestor": ingestor.Name(), "error": err}).Fatal()
+		}
+
+		doc.Find(".node-project-module").Each(func(i int, s *goquery.Selection) {
+			if !done {
+				var id string
+				if idAttr, exists := s.Attr("id"); exists {
+					parts := strings.SplitN(idAttr, "-", 2) // e.g. "node-1234"
+					id = parts[1]
+				}
+				packageResults := ingestor.getVersions(id)
+				if len(packageResults) == 0 { // last page didnt' have any new versions, so break and return all collected versions
+					done = true
+				} else {
+					results = append(results, packageResults...)
+				}
+			}
+		})
+		page++
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	ingestor.LatestRun = time.Now()
+
+	return results
+}
+
+func (ingestor *Drupal) getVersions(id string) []data.PackageVersion {
+	var results []data.PackageVersion
+	fp := gofeed.NewParser()
+
+	feed, err := fp.ParseURL(fmt.Sprintf(drupalReleasesUrl, id))
+	if err != nil {
+		log.WithFields(log.Fields{"ingestor": ingestor.Name()}).Error(err)
+		return results
+	}
+
+	for _, item := range feed.Items { // Mon Jan 2 15:04:05 -0700 MST 2006
+		createdAtTime, _ := time.Parse(time.RFC1123, item.Published)
+		nameAndVersion := strings.SplitN(item.Title, " ", 2) // e.g. ctools 7.x-1.19
```

Comment:
Libraries is going to have to handle these versions once it gets them and turn them into regular versions -- it may be smart to only support Drupal 8.x versions too, since it adopted semver: https://www.drupal.org/docs/8/understanding-drupal-version-numbers/what-do-version-numbers-mean-on-contributed-modules-and